feat: allow entities to be included multiple times

mKeRix opened the pull request:
This PR adds a "plus-addressing" feature similar to what Gmail does in order to allow the same HA entity id to be used multiple times in YIO. This is useful for controlling multiple devices/setups through the same remote entity in HA for example. As discussed on Discord "remote.harmony_hub" could now be referenced as "remote.harmony_hub+appletv" or "remote.harmony_hub+shield" within YIO, with the references being resolved correctly. For consistency reasons this change applies to all entity types in Home Assistant even though the main use case is the remote type. Entity IDs without a `+` will still function as before, this is strictly a feature addition.

I've tested this with multiple entity types, checking that all entries can be controlled correctly and updates in HA are reflected across all.

**Example config:**
```json
{
    "entities": {
        "remote": [
            {
                "commands": [
                    {
                        "button_map": "CURSOR_UP",
                        "device": "1234",
                        "code": "up"
                    }
                ],
                "entity_id": "remote.harmony_hub+shield",
                "friendly_name": "NVIDIA Shield",
                "integration": "someid",
                "supported_features": [
                    "CURSOR_UP"
                ]
            },
            {
                "commands": [
                    {
                        "button_map": "CURSOR_UP",
                        "device": "6789",
                        "code": "up"
                    }
                ],
                "entity_id": "remote.harmony_hub+philipstv",
                "friendly_name": "Philips TV",
                "integration": "someid",
                "supported_features": [
                    "CURSOR_UP"
                ]
            }
        ],
    }
}
```

mKeRix committed:
feat: allow entities to be included multiple times

This PR adds a "plus-addressing" feature similar to what Gmail does in order to allow the same HA entity id to be used multiple times in YIO. This is useful for controlling multiple devices/setups through the same remote entity in HA for example. As discussed in Discord.

mKeRix committed:
chore: fix code style

martonborzak commented:
Thanks @mKeRix! This will only work with manual configuration. When the integration asks Home Assistant to list all available entities, it won't give you the `+appletv` part. That needs to be changed manually afterwards. We need to think about that in the future, but this will work now I think.

mKeRix commented:
Yep, I just wanted to deliver a first iteration of the feature. Configuring a remote with HA can't be done without touching the JSON manually anyway atm. An idea to add this to the web configurator without starting to build very integration-specific into it: have an advanced mode which allows you to customize the entity id (or maybe just the JSON in general?) when adding an item.

zehnm commented on src/homeassistant.cpp:
```diff
@@ -295,22 +295,25 @@ void HomeAssistant::webSocketSendCommand(const QString &domain, const QString &s
-        }
-        if (entity->type() == "switch") {
-            updateSwitch(entity, attr);
+    QList<EntityInterface*> entities = m_entities->getByIntegration(integrationId());
+
+    foreach(EntityInterface *entity, entities) {
+        if (entity->entity_id().startsWith(entity_id)) {
```
I'm not that thrilled about parsing all entities until we have a match. Compared to the previous optimized single map access, this is a huge overhead. Especially because it applies to all entity types and the situation gets worse the more entities we have.

Wasn't this change only intended for the remote entity?
If this is the case, then there's no reason to change `updateEntity` since there's no remote entity update. Or I miss something...

FYI: Qt's `foreach` macro is deprecated and will be removed in a future version of Qt. See: https://doc.qt.io/qt-5/qtglobal.html#foreach 
Just be careful when using the new C++11 range-for with Qt containers, so there's no deep copy of the container!

zehnm reviewed (changes requested):
@martonborzak @mKeRix I'd like to discuss this PR further, before allowing a potentially performance degrading change being merged.
E.g. 
- Do we really need this addressing feature for all entities?
  What are the use cases besides the remote entity?
- How does this affect the Homey and OpenHAB integrations?

mKeRix commented:
Thanks for the review @zehnm! As mentioned in the original post the decision to implement this for all entities was purely for consistency reasons, I can't think of a reason to use this with any other supported entity type atm. My assumption there was that on hardware like a Pi Zero the performance impact wouldn't really be noticeable. I'm happy to revise this PR and limit the feature to remote entities only again, which would let me revert the update method back to the direct map access again.

Note that there are state updates for remote entities - they have a binary state ("on"/"off") in HA. I don't think they are used by YIO in any way though, so I suppose it doesn't matter.

mKeRix committed:
fix: limit plus addressing to remotes

As requested in review so that we don't impact performance for a feature that isn't really useful for non-remote entities atm.

mKeRix commented:
I adapted it to only work for remote entities, which makes the PR a lot smaller. I tested it again for regular entities and multiple reference to the same remote entity, works fine for me.

zehnm reviewed (approved):
Thanks for limiting it to the remote entity only!
